components/suggestion/suggestion_box/suggestion_box.tsx

```diff
@@ -0,0 +1,776 @@
+        }
+    }, [containerRef, onFocus, setState, pretext, preventSuggestionListCloseFlag, inputRef, openOnFocus, openWhenEmpty, requiredCharacters]);
+
+    const handleKeyDown = (e: KeyboardEvent<HTMLDivElement | TextboxElement>) => {
+        if ((openWhenEmpty || value) && hasSuggestions(state.items)) {
+            const ctrlOrMetaKeyPressed = e.ctrlKey || e.metaKey;
+            if (Utils.isKeyPressed(e, KeyCodes.UP)) {
+                selectPrevious(state.terms, state.selection, setState);
+                e.preventDefault();
+            } else if (Utils.isKeyPressed(e, KeyCodes.DOWN)) {
+                selectNext(state.terms, state.selection, setState);
+                e.preventDefault();
+            } else if (
+                (Utils.isKeyPressed(e, KeyCodes.ENTER) && !ctrlOrMetaKeyPressed) ||
+                (completeOnTab && Utils.isKeyPressed(e, KeyCodes.TAB))
+            ) {
+                let matchedPretext = '';
+                for (let i = 0; i < state.terms.length; i++) {
+                    if (state.terms[i] === state.selection) {
+                        matchedPretext = state.matchedPretext[i];
+                    }
+                }
+
+                // If these don't match, the user typed quickly and pressed enter before we could
+                // update the pretext, so update the pretext before completing
+                if (pretext.toLowerCase().endsWith(matchedPretext.toLowerCase())) {
+                    if (handleCompleteWord(state.selection, matchedPretext, e)) {
+                        return;
+                    }
+                } else {
+                    if (timeoutId) {
+                        clearTimeout(timeoutId);
+                    }
+                    nonDebouncedPretextChanged(
+                        pretext,
+                        true,
+                        providers,
+                        state,
+                        inputRef,
+                        setState,
+                        (suggestions: ProviderSuggestions) => handleReceivedSuggestions(suggestions, setState, state, onSuggestionsReceived),
+                        (suggestions: ProviderSuggestions) => handleReceivedSuggestionsAndComplete(suggestions, setState, state, handleCompleteWord, onSuggestionsReceived),
+                    );
+                }
+
+                if (onKeyDown) {
+                    onKeyDown(e);
+                }
+                e.preventDefault();
+            } else if (Utils.isKeyPressed(e, KeyCodes.ESCAPE)) {
+                clear(setState, state);
+                setState((prevState) => ({...prevState, presentationType: 'text'}));
+                e.preventDefault();
+                e.stopPropagation();
+            } else if (onKeyDown) {
+                onKeyDown(e);
+            }
+        } else if (onKeyDown) {
+            onKeyDown(e);
+        }
+    };
+
+    useEffect(() => {
+        handlePretextChanged(pretext);
+    }, [pretext]);
```

Comment:
ESlint wants this dep array to include `handlePretextChanged`. If you do that, then it goes into a rerender loop because pretext changes on `handlePretextChanged`. If you have any suggestions on this one.

Reply:
It's not actually going into a render loop because of that. It's getting called because `handlePretextChanged` is declared in this function, it's re-declared every time this function is called, so it changes on every single render and the `useEffect` triggers every render because of that.

We can probably get rid of `handlePretextChanged` entirely and called `debouncedPretextChanged` in here to solve that issue. There might be other props that need to be passed in as well then, but hopefully none of those will cause similar issues.

I almost wonder if we can get rid of `pretext` in the state entirely, but that's a much bigger change